### PR TITLE
Integrate with quota parser

### PR DESCRIPTION
**Pull request by qiwzhang**
* Use RequestContext to store quota vector
* Add cache map for ServiceContext so it can be used to preprocess config data.
* Change quota ConfigParser GetRequirements() function to not to return quota vector, instead pass its raw pointer as last argument so the data can be appended.

### CODE REVIEW

File: control/src/http/client_context.h [hunk 1]
```diff
@@ -47,8 +47,31 @@ class ClientContext : public ClientContextBase {
...
-    return legacy_quotas_;
+  // Append the legacy quotas.
+  void AddLegacyQuotas(std::vector<::istio::quota::Requirement>* quotas) const {
+    quotas->insert(quotas->end(), legacy_quotas_.begin(), legacy_quotas_.end());
+  }
+
+  // Get valid service name in the config map.
+  // If input service name is in the map, use it, otherwise, use the default
+  // one.
+  const std::string& GetServiceName(const std::string& service_name) const {
```
Comment (ayj) [inline control/src/http/client_context.h]: Is there ever a case where `GetServiceName` and `GetServiceConfig` would not be used together? As it is right now, this is doing a duplicate map look-up back-to-back which could be combined.

Comment (qiwzhang) [reply control/src/http/client_context.h]: The first call result is used for checking the cache map.   If cache hit, the second call is not needed. 
But you are right,  it is two map lookup; first is the service_configs map,  the  second is the cache service_context_map_.  I will optimize it to one lookup.

File: control/src/http/controller_impl.cc [hunk 2]
```diff
@@ -39,14 +39,21 @@ std::shared_ptr<ServiceContext> ControllerImpl::GetServiceContext(
     return std::make_shared<ServiceContext>(client_context_,
                                             *config.legacy_config);
   }
-  auto config_map = client_context_->config().service_configs();
-  auto it = config_map.find(config.destination_service);
-  if (it == config_map.end()) {
-    it = config_map.find(
-        client_context_->config().default_destination_service());
+
+  const std::string& origin_name = config.destination_service;
```
Comment (ayj) [inline control/src/http/controller_impl.cc]: Where is cache cleared? Do we just create a new controller any time the client config changes?

Comment (qiwzhang) [reply control/src/http/controller_impl.cc]: Yes.  if client config changed,  xDS will send new envoy filter config, and new controller will be created.

**Pull request merged**